{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.016615,
     "end_time": "2020-12-16T19:23:01.492115",
     "exception": false,
     "start_time": "2020-12-16T19:23:01.475500",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Using Shotgun metagenomics for disease prediction\n",
    "Shotgun metagenomic sequencing is a relatively new sequencing approach that allows insight to be gained into community biodiversity and function.<br>\n",
    "The function of shotgun metagenomic sequencing is to sequence the genomes of untargeted cells in a community in order to elucidate community composition and function.<br>\n",
    "In this notebook I depart from the discoveries made in the [dataset's creators research paper](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004977), and therefore skip EDA. I wanted to expand on their work and first of all I approach prediction cirrhosis using XGBoost in 'synthetic pure' case of samples with cirrhosis against healthy samples. Creating binary classification problem this way, and comparing it with the result they achieved using RandomForest and SVM. <br>\n",
    "I'm also checking if the model trained on this dataset would be still able to make prediction if we given a dataset with other diseases mixed in. <br>\n",
    "And finally I want to check if the model will be able to distinguish cirrhosis from other diseases."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.013189,
     "end_time": "2020-12-16T19:23:01.519067",
     "exception": false,
     "start_time": "2020-12-16T19:23:01.505878",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "![](https://d3i71xaburhd42.cloudfront.net/fb32dc53b3f5c638d84fb02da7213863a27cee28/2-Figure1-1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.013207,
     "end_time": "2020-12-16T19:23:01.546093",
     "exception": false,
     "start_time": "2020-12-16T19:23:01.532886",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "\n",
    "A few useful articles:<br>\n",
    "[An introduction to the analysis of shotgun metagenomic data](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4059276/)<br>\n",
    "[Shotgun metagenomics, from sampling to analysis](https://pubmed.ncbi.nlm.nih.gov/28898207/)<br>\n",
    "[16S rRNA Gene Sequencing for Bacterial Identification in the Diagnostic Laboratory: Pluses, Perils, and Pitfalls](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2045242/)<br>\n",
    "[Then and now: use of 16S rDNA gene sequencing for bacterial identification and discovery of novel bacteria in clinical microbiology laboratories](https://pubmed.ncbi.nlm.nih.gov/18828852/)<br>\n",
    "[Direct 16S rRNA-seq from bacterial communities: a PCR-independent approach to simultaneously assess microbial diversity and functional activity potential of each taxon](https://www.nature.com/articles/srep32165#abstract)<br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.013219,
     "end_time": "2020-12-16T19:23:01.572643",
     "exception": false,
     "start_time": "2020-12-16T19:23:01.559424",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Loading libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
    "_kg_hide-input": true,
    "_kg_hide-output": true,
    "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5",
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:01.609303Z",
     "iopub.status.busy": "2020-12-16T19:23:01.608576Z",
     "iopub.status.idle": "2020-12-16T19:23:03.950988Z",
     "shell.execute_reply": "2020-12-16T19:23:03.950355Z"
    },
    "papermill": {
     "duration": 2.365116,
     "end_time": "2020-12-16T19:23:03.951113",
     "exception": false,
     "start_time": "2020-12-16T19:23:01.585997",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/kaggle/input/human-metagenomics/markers2clades_DB.csv\n",
      "/kaggle/input/human-metagenomics/abundance.csv\n",
      "/kaggle/input/human-metagenomics/abundance_stoolsubset.csv\n",
      "/kaggle/input/metagenomics/markers2clades_DB.txt\n",
      "/kaggle/input/metagenomics/abundance_stoolsubset.txt\n",
      "/kaggle/input/metagenomics/abundance.txt\n",
      "/kaggle/input/metagenomics/marker_presence.txt\n"
     ]
    }
   ],
   "source": [
    "# This Python 3 environment comes with many helpful analytics libraries installed\n",
    "# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python\n",
    "# For example, here's several helpful packages to load\n",
    "\n",
    "import numpy as np # linear algebra\n",
    "import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)\n",
    "\n",
    "# Input data files are available in the read-only \"../input/\" directory\n",
    "# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory\n",
    "\n",
    "import os\n",
    "for dirname, _, filenames in os.walk('/kaggle/input'):\n",
    "    for filename in filenames:\n",
    "        print(os.path.join(dirname, filename))\n",
    "\n",
    "from xgboost import XGBClassifier\n",
    "import xgboost as xgb\n",
    "\n",
    "import sklearn\n",
    "from sklearn import metrics\n",
    "from sklearn import preprocessing\n",
    "from sklearn.model_selection import StratifiedKFold\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import f1_score, roc_auc_score, confusion_matrix, precision_recall_curve, auc, roc_curve, recall_score\n",
    "\n",
    "import seaborn as sns\n",
    "import scikitplot as skplt\n",
    "import matplotlib.pyplot as plt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "_cell_guid": "79c7e3d0-c299-4dcb-8224-4455121ee9b0",
    "_uuid": "d629ff2d2480ee46fbb7e2d37f6b5fab8052498a",
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:03.985615Z",
     "iopub.status.busy": "2020-12-16T19:23:03.984928Z",
     "iopub.status.idle": "2020-12-16T19:23:03.988324Z",
     "shell.execute_reply": "2020-12-16T19:23:03.987591Z"
    },
    "papermill": {
     "duration": 0.022494,
     "end_time": "2020-12-16T19:23:03.988437",
     "exception": false,
     "start_time": "2020-12-16T19:23:03.965943",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Metaparameters\n",
    "VERBOSE = 0\n",
    "FOLDS = 5\n",
    "\n",
    "show_fold_stats = True\n",
    "# show_fold_stats = False # set to True if all OOF results wanted\n",
    "\n",
    "do_plot_ROC = False # set to True to plot ROC when predicting cirrhosis\n",
    "\n",
    "# test_train_split_SEED = 1970\n",
    "test_train_split_SEED = 1971"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "_kg_hide-input": true,
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:04.029315Z",
     "iopub.status.busy": "2020-12-16T19:23:04.028549Z",
     "iopub.status.idle": "2020-12-16T19:23:04.032017Z",
     "shell.execute_reply": "2020-12-16T19:23:04.031382Z"
    },
    "papermill": {
     "duration": 0.029628,
     "end_time": "2020-12-16T19:23:04.032130",
     "exception": false,
     "start_time": "2020-12-16T19:23:04.002502",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def plot_ROC(fpr, tpr, m_name):\n",
    "    roc_auc = sklearn.metrics.auc(fpr, tpr)\n",
    "    plt.figure(figsize=(6, 6))\n",
    "    lw = 2\n",
    "    plt.plot(fpr, tpr, color='darkorange',\n",
    "             lw=lw, label='ROC curve (area = %0.2f)' % roc_auc, alpha=0.5)\n",
    "    \n",
    "    plt.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--', alpha=0.5)\n",
    "    \n",
    "    plt.xlim([0.0, 1.0])\n",
    "    plt.ylim([0.0, 1.05])\n",
    "    plt.xticks(fontsize=16)\n",
    "    plt.yticks(fontsize=16)\n",
    "    plt.grid(True)\n",
    "    plt.xlabel('False Positive Rate', fontsize=16)\n",
    "    plt.ylabel('True Positive Rate', fontsize=16)\n",
    "    plt.title('ROC for %s'%m_name, fontsize=20)\n",
    "    plt.legend(loc=\"lower right\", fontsize=16)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.013919,
     "end_time": "2020-12-16T19:23:04.060304",
     "exception": false,
     "start_time": "2020-12-16T19:23:04.046385",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "My initial experiments with full dataset of species abundance features proved that XGB wasn't able to beat the original paper results. So I benefited from their analyses of features importance, and used the reduced dataset for ongoing experimentation. <br>\n",
    "I still plan on using markers in future experiments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "_kg_hide-output": true,
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:04.093707Z",
     "iopub.status.busy": "2020-12-16T19:23:04.092721Z",
     "iopub.status.idle": "2020-12-16T19:23:05.594248Z",
     "shell.execute_reply": "2020-12-16T19:23:05.593223Z"
    },
    "papermill": {
     "duration": 1.519013,
     "end_time": "2020-12-16T19:23:05.594399",
     "exception": false,
     "start_time": "2020-12-16T19:23:04.075386",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n                             944\n",
      "t2d                           223\n",
      "obesity                       164\n",
      "ibd_ulcerative_colitis        148\n",
      "cirrhosis                     118\n",
      "leaness                        89\n",
      "stec2-positive                 52\n",
      "impaired_glucose_tolerance     49\n",
      "cancer                         48\n",
      "n_relative                     47\n",
      "small_adenoma                  26\n",
      "ibd_crohn_disease              25\n",
      " -                             20\n",
      "large_adenoma                  13\n",
      "overweight                     10\n",
      "-                               7\n",
      "obese                           5\n",
      "underweight                     1\n",
      "Name: disease, dtype: int64\n",
      "Neilsen_genome_assembly              382\n",
      "t2dmeta_long                         290\n",
      "Chatelier_gut_obesity                278\n",
      "Quin_gut_liver_cirrhosis             232\n",
      "hmp                                  152\n",
      "WT2D                                 145\n",
      "Zeller_fecal_colorectal_cancer       134\n",
      "metahit                              110\n",
      "t2dmeta_short                         73\n",
      "Tito_subsistence_gut                  58\n",
      "hmpii                                 44\n",
      "Loman2013_EcoliOutbreak_DNA_HiSeq     44\n",
      "Candela_Africa                        38\n",
      "Loman2013_EcoliOutbreak_DNA_MiSeq      9\n",
      "Name: dataset_name, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# pd_abundance = pd.read_csv('/kaggle/input/metagenomics/abundance.txt', sep='\\t', header=None, index_col=0, dtype=str)\n",
    "pd_abundance = pd.read_csv('../input/human-metagenomics/abundance_stoolsubset.csv', dtype=str)\n",
    "disease = pd_abundance.loc[:,'disease'] \n",
    "d_name = pd_abundance.loc[:,'dataset_name'] \n",
    "print(disease.value_counts())\n",
    "print(d_name.value_counts())\n",
    "\n",
    "# list of diseases we want to analyze and predict\n",
    "diseases = ['obesity', 'cirrhosis', 't2d', 'cancer']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:05.700941Z",
     "iopub.status.busy": "2020-12-16T19:23:05.699812Z",
     "iopub.status.idle": "2020-12-16T19:23:06.535306Z",
     "shell.execute_reply": "2020-12-16T19:23:06.533919Z"
    },
    "papermill": {
     "duration": 0.925786,
     "end_time": "2020-12-16T19:23:06.535436",
     "exception": false,
     "start_time": "2020-12-16T19:23:05.609650",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "cols = pd_abundance.columns.tolist()\n",
    "species = [x for x in cols if x.startswith('k_')]\n",
    "metadata = [x for x in cols if not x.startswith('k_')]\n",
    "pd_abundance_conv = pd_abundance.copy()\n",
    "pd_abundance_conv = pd_abundance_conv[species].astype('float64')\n",
    "pd_abundance_conv = pd.concat([pd_abundance[metadata], pd_abundance_conv], axis = 1)\n",
    "\n",
    "# controls/healthy samples from Human Microbiome Project coded 'hmp' and 'hmpii'. \n",
    "# 't2d' stands for Type 2 Diabetes. We will combine a few studies into single dataset.\n",
    "data_sets = {'control':['hmp', 'hmpii'],'t2d':['WT2D','t2dmeta_long','t2dmeta_short'], 'cirrhosis' : ['Quin_gut_liver_cirrhosis'], \n",
    "             'cancer' : ['Zeller_fecal_colorectal_cancer'], 'obesity' : ['Chatelier_gut_obesity']}\n",
    "# combine controls from different studies into one\n",
    "pd_abundance_conv['disease'] = pd_abundance_conv['disease'].apply(lambda x: 'control' if ((x == 'n') or (x == 'nd') or (x == 'leaness')) else x)\n",
    "# separate controls and diseases into 2 dataframes\n",
    "pd_control = pd_abundance_conv.loc[pd_abundance_conv['disease'] == 'control']\n",
    "pd_disease = pd_abundance_conv.loc[pd_abundance_conv['disease'] != 'control']\n",
    "\n",
    "# we won't consider diseases from this list\n",
    "not_disease = [d for d in pd_disease.disease.unique().tolist() if d not in diseases] \n",
    "for d in not_disease:\n",
    "    pd_disease = pd_disease.drop(pd_disease.loc[pd_disease['disease'] == d].index, axis = 0)     "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.015017,
     "end_time": "2020-12-16T19:23:06.565521",
     "exception": false,
     "start_time": "2020-12-16T19:23:06.550504",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## Defining model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "_kg_hide-input": true,
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:06.600902Z",
     "iopub.status.busy": "2020-12-16T19:23:06.600104Z",
     "iopub.status.idle": "2020-12-16T19:23:06.686562Z",
     "shell.execute_reply": "2020-12-16T19:23:06.687273Z"
    },
    "papermill": {
     "duration": 0.107046,
     "end_time": "2020-12-16T19:23:06.687451",
     "exception": false,
     "start_time": "2020-12-16T19:23:06.580405",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def disease_distinction(d, counter_d):\n",
    "    print('-' * 80)\n",
    "    print('Discriminate %s from %s'%(d,counter_d))\n",
    "    \n",
    "    skf = StratifiedKFold(n_splits = FOLDS, shuffle = True, random_state = SEED)\n",
    "    oof_preds = []\n",
    "    oof_aucs = []\n",
    "    oof_scores= []\n",
    "    ds_names = data_sets[d]\n",
    "    if len(ds_names) == 1:\n",
    "        pd_cont = pd_control.loc[pd_control['dataset_name'] == ds_names[0]]\n",
    "        pd_dis = pd_disease.loc[pd_disease['dataset_name'] == ds_names[0]]\n",
    "    else:\n",
    "        pd_cont = pd_control.loc[pd_control['dataset_name'] == ds_names[0]]\n",
    "        pd_dis = pd_disease.loc[(pd_disease['disease'] == d) & (pd_disease['dataset_name'] == ds_names[0])]\n",
    "        for ds in ds_names[1:]:\n",
    "            pd_cont = pd.concat([pd_cont, pd_control.loc[pd_control['dataset_name'] == ds]], axis = 0)\n",
    "            pd_dis = pd.concat([pd_dis, pd_disease.loc[(pd_disease['disease'] == d) & (pd_disease['dataset_name'] == ds)]], axis = 0)\n",
    "                                 \n",
    "    #     create dataset with all other diseases, target set to 0\n",
    "    pd_others = pd_disease.loc[(pd_disease['disease'] == counter_d)]\n",
    "    target_others = pd_others['disease'].apply(lambda x: 1 if x == d else 0)\n",
    "    \n",
    "    pd_train = pd.concat([pd_cont, pd_dis], axis = 0)\n",
    "    #     adding control data from healthy subject, data by HMP\n",
    "    pd_train = pd.concat([pd_train, pd_control.loc[pd_control['dataset_name'] == 'hmp']], axis = 0)\n",
    "    pd_train = pd.concat([pd_train, pd_control.loc[pd_control['dataset_name'] == 'hmpii']], axis = 0)\n",
    "    target = pd_train['disease']\n",
    "    #     convert text target into binary\n",
    "    binary_target = target.apply(lambda x: 1 if x == d else 0)\n",
    "    # binary_target.value_counts()\n",
    "    #     use only abundance data for training\n",
    "    pd_others = pd_others[species] \n",
    "    pd_train  = pd_train[species] \n",
    "\n",
    "    #     this split provides us with preserved test set\n",
    "    disease_train, disease_test, disease_y_train, disease_y_test = train_test_split(pd_train, binary_target, test_size = 0.10, \n",
    "#                                                                                     stratify = binary_target,\n",
    "                                                                                    random_state = test_train_split_SEED)   \n",
    "\n",
    "    #     combining preserved test set with othere diseases samples\n",
    "    full_test = pd.concat([disease_test, pd_others])\n",
    "    full_y_test = pd.concat([disease_y_test, target_others])\n",
    "    \n",
    "    disease_y_test.value_counts()\n",
    "    disease_y_train.value_counts()\n",
    "    preds = np.zeros(disease_y_test.shape[0])\n",
    "    full_preds = np.zeros(full_y_test.shape[0])\n",
    "\n",
    "    for fold, (idxT,idxV) in enumerate(skf.split(disease_train, disease_y_train)):\n",
    "\n",
    "        X_train = disease_train.iloc[idxT]\n",
    "        X_val = disease_train.iloc[idxV]\n",
    "        y_train = disease_y_train.iloc[idxT]\n",
    "        y_val = disease_y_train.iloc[idxV]\n",
    "    \n",
    "        XGB_model = XGBClassifier(n_estimators=5000, max_depth=None, \n",
    "                            learning_rate=0.05,\n",
    "                            objective='binary:logistic', \n",
    "                            metric='auc',\n",
    "                            verbosity  = VERBOSE,\n",
    "                            n_jobs=-1, random_state  = SEED)\n",
    "        \n",
    "        if show_fold_stats:\n",
    "            print('-' * 80)\n",
    "            print('Fold : %s'%(fold+1))\n",
    "    \n",
    "        XGB_model.fit(X_train, y_train,\n",
    "                        eval_set = [(X_val, y_val)],\n",
    "                        eval_metric=['logloss'],\n",
    "                        early_stopping_rounds = 100, verbose = VERBOSE )\n",
    "            \n",
    "        XGB_preds = XGB_model.predict_proba(X_val)\n",
    "        XGB_score = metrics.roc_auc_score(y_val, XGB_preds[:,1])\n",
    "        XGB_class = XGB_model.predict(X_val)\n",
    "        \n",
    "        XGB_test = XGB_model.predict_proba(disease_test)\n",
    "        XGB_test_score = metrics.roc_auc_score(disease_y_test, XGB_test[:,1])\n",
    "        XGB_test_class = XGB_model.predict(disease_test)\n",
    "        \n",
    "        full_test_preds = XGB_model.predict_proba(full_test)\n",
    "        full_test_score = metrics.roc_auc_score(full_y_test, full_test_preds[:,1])\n",
    "        full_test_class = XGB_model.predict(full_test)\n",
    "               \n",
    "        f1s = f1_score(y_val, XGB_class)\n",
    "        recall = metrics.recall_score(y_val, XGB_class)\n",
    "        precision_score = metrics.precision_score(y_val, XGB_class)\n",
    "        \n",
    "        f1_test = f1_score(disease_y_test, XGB_test_class)\n",
    "        recall_test = metrics.recall_score(disease_y_test, XGB_test_class)\n",
    "        precision_score_test = metrics.precision_score(disease_y_test, XGB_test_class)\n",
    "        \n",
    "        f1_full_test = f1_score(full_y_test, full_test_class)\n",
    "        recall_full_test = metrics.recall_score(full_y_test, full_test_class)\n",
    "        precision_full_test = metrics.precision_score(full_y_test, full_test_class)\n",
    "\n",
    "        if show_fold_stats:        \n",
    "            print('ROC AUC score for XGBoost model valid set: %.4f'%XGB_score)\n",
    "            print('F1 score: %0.4f'%f1s)\n",
    "            print(confusion_matrix(y_val, XGB_class))\n",
    "    \n",
    "            print('ROC AUC score for XGBoost model test set: %.4f'%XGB_test_score)\n",
    "            print('F1 score: %0.4f'%f1_test)\n",
    "            print(confusion_matrix(disease_y_test, XGB_test_class))\n",
    "            \n",
    "            print('ROC AUC score for full test set including other diseases used as false controls: %.4f'%full_test_score)\n",
    "            print('F1 score: %0.4f'%f1_full_test)\n",
    "            print(confusion_matrix(full_y_test, full_test_class))\n",
    "        \n",
    "        preds += XGB_test[:,1] / FOLDS\n",
    "        full_preds += full_test_preds[:,1] / FOLDS\n",
    "        \n",
    "        fold_score = [XGB_score,f1s,recall,precision_score, XGB_test_score,f1_test,recall_test,precision_score_test]\n",
    "        oof_scores.append({fold : fold_score})\n",
    "    \n",
    "    avg_test_score = metrics.roc_auc_score(disease_y_test, preds)\n",
    "    avg_class = np.where(preds < 0.5, 0, 1)\n",
    "    avg_f1_test = f1_score(disease_y_test, avg_class)\n",
    "    avg_recall_test = metrics.recall_score(disease_y_test, avg_class)\n",
    "    avg_precision_score_test = metrics.precision_score(disease_y_test, avg_class)\n",
    "\n",
    "    avg_full_test_score = metrics.roc_auc_score(full_y_test, full_preds)\n",
    "    avg_class_full = np.where(full_preds < 0.5, 0, 1)\n",
    "    avg_f1_test_full = f1_score(full_y_test, avg_class_full)\n",
    "    avg_recall_full_test = metrics.recall_score(full_y_test, avg_class_full)\n",
    "    avg_precision_full_test = metrics.precision_score(full_y_test, avg_class_full)\n",
    "\n",
    "    if show_fold_stats:        \n",
    "        print('-' * 80)\n",
    "        print('ROC AUC score for %s averaged over %i folds: %.4f'%(d, FOLDS, avg_test_score))\n",
    "        print('F1 : %.4f, Recall : %.4f , Precision : %.4f'%(avg_f1_test, avg_recall_test, avg_precision_score_test))\n",
    "        print('Confusion matrix for %s averaged across %i folds '%(d,FOLDS))\n",
    "        print(confusion_matrix(disease_y_test, avg_class))\n",
    "\n",
    "    \n",
    "    print('ROC AUC score for %s against %s averaged over %i folds : %.4f'%(d, counter_d, FOLDS, avg_full_test_score ))\n",
    "    print('F1 : %.4f, Recall : %.4f , Precision : %.4f'%(avg_f1_test_full, avg_recall_full_test, avg_precision_full_test))\n",
    "    print('Confusion matrix for %s against %s averaged across %i folds'%(d, counter_d, FOLDS))\n",
    "    print(confusion_matrix(full_y_test, avg_class_full))\n",
    "    \n",
    "    return full_preds, full_y_test, {d : (avg_test_score, avg_f1_test, avg_recall_test, avg_precision_score_test, oof_scores)}\n",
    "\n",
    "\n",
    "def disease_prediction(d):\n",
    "    print('-' * 80)\n",
    "    print('Disease : %s'%d)\n",
    "    \n",
    "    skf = StratifiedKFold(n_splits = FOLDS, shuffle = True, random_state = SEED)\n",
    "    oof_preds = []\n",
    "    oof_aucs = []\n",
    "    oof_scores= []\n",
    "    ds_names = data_sets[d]\n",
    "    if len(ds_names) == 1:\n",
    "        pd_cont = pd_control.loc[pd_control['dataset_name'] == ds_names[0]]\n",
    "        pd_dis = pd_disease.loc[pd_disease['dataset_name'] == ds_names[0]]\n",
    "    else:\n",
    "        pd_cont = pd_control.loc[pd_control['dataset_name'] == ds_names[0]]\n",
    "        pd_dis = pd_disease.loc[(pd_disease['disease'] == d) & (pd_disease['dataset_name'] == ds_names[0])]\n",
    "        for ds in ds_names[1:]:\n",
    "            pd_cont = pd.concat([pd_cont, pd_control.loc[pd_control['dataset_name'] == ds]], axis = 0)\n",
    "            pd_dis = pd.concat([pd_dis, pd_disease.loc[(pd_disease['disease'] == d) & (pd_disease['dataset_name'] == ds)]], axis = 0)\n",
    "                \n",
    "    #     create dataset with all other diseases, target set to 0\n",
    "    pd_others = pd_disease.loc[(pd_disease['disease'] != d)]\n",
    "    target_others = pd_others['disease'].apply(lambda x: 1 if x == d else 0)\n",
    "    \n",
    "    #     combine controls and this particular disease back into train dataset\n",
    "    pd_train = pd.concat([pd_cont, pd_dis], axis = 0)\n",
    "\n",
    "    #     adding control data from healthy subject, data by HMP\n",
    "    pd_train = pd.concat([pd_train, pd_control.loc[pd_control['dataset_name'] == 'hmp']], axis = 0)\n",
    "    pd_train = pd.concat([pd_train, pd_control.loc[pd_control['dataset_name'] == 'hmpii']], axis = 0)\n",
    "\n",
    "    target = pd_train['disease']\n",
    "    #     convert text target into binary\n",
    "    binary_target = target.apply(lambda x: 1 if x == d else 0)\n",
    "    # binary_target.value_counts()\n",
    "    #     use only abundance features data for training\n",
    "    pd_others = pd_others[species] \n",
    "    pd_train  = pd_train[species] \n",
    "    #     this split provides us with preserved test set\n",
    "    disease_train, disease_test, disease_y_train, disease_y_test = train_test_split(pd_train, binary_target, test_size = 0.10, \n",
    "#                                                                                     stratify = binary_target,\n",
    "                                                                                    random_state = test_train_split_SEED)   \n",
    "        \n",
    "    #     combining preserved test set with othere diseases samples\n",
    "    full_test = pd.concat([disease_test, pd_others])\n",
    "    full_y_test = pd.concat([disease_y_test, target_others])\n",
    "    \n",
    "    disease_y_test.value_counts()\n",
    "    disease_y_train.value_counts()\n",
    "    preds = np.zeros(disease_y_test.shape[0])\n",
    "    full_preds = np.zeros(full_y_test.shape[0])\n",
    "\n",
    "    for fold, (idxT,idxV) in enumerate(skf.split(disease_train, disease_y_train)):\n",
    "\n",
    "        X_train = disease_train.iloc[idxT]\n",
    "        X_val = disease_train.iloc[idxV]\n",
    "        y_train = disease_y_train.iloc[idxT]\n",
    "        y_val = disease_y_train.iloc[idxV]\n",
    "    \n",
    "        XGB_model = XGBClassifier(n_estimators=5000, max_depth=None, \n",
    "                            learning_rate=0.05,\n",
    "                            objective='binary:logistic', \n",
    "                            metric='auc',\n",
    "                            verbosity  = VERBOSE,\n",
    "                            n_jobs=-1, random_state  = SEED)\n",
    "        \n",
    "        if show_fold_stats:\n",
    "            print('-' * 80)\n",
    "            print('Fold : %s'%(fold+1))\n",
    "    \n",
    "        XGB_model.fit(X_train, y_train,\n",
    "                        eval_set = [(X_val, y_val)],\n",
    "                        eval_metric=['logloss'],\n",
    "                        # eval_metric=['auc','logloss'],\n",
    "                        early_stopping_rounds = 100, verbose = VERBOSE )\n",
    "            \n",
    "        XGB_preds = XGB_model.predict_proba(X_val)\n",
    "        XGB_score = metrics.roc_auc_score(y_val, XGB_preds[:,1])\n",
    "        XGB_class = XGB_model.predict(X_val)\n",
    "        \n",
    "        XGB_test = XGB_model.predict_proba(disease_test)\n",
    "        XGB_test_score = metrics.roc_auc_score(disease_y_test, XGB_test[:,1])\n",
    "        XGB_test_class = XGB_model.predict(disease_test)\n",
    "        \n",
    "        full_test_preds = XGB_model.predict_proba(full_test)\n",
    "        full_test_score = metrics.roc_auc_score(full_y_test, full_test_preds[:,1])\n",
    "        full_test_class = XGB_model.predict(full_test)\n",
    "              \n",
    "        f1s = f1_score(y_val, XGB_class)\n",
    "        recall = metrics.recall_score(y_val, XGB_class)\n",
    "        precision_score = metrics.precision_score(y_val, XGB_class)\n",
    "        \n",
    "        f1_test = f1_score(disease_y_test, XGB_test_class)\n",
    "        recall_test = metrics.recall_score(disease_y_test, XGB_test_class)\n",
    "        precision_score_test = metrics.precision_score(disease_y_test, XGB_test_class)\n",
    "        \n",
    "        f1_full_test = f1_score(full_y_test, full_test_class)\n",
    "        recall_full_test = metrics.recall_score(full_y_test, full_test_class)\n",
    "        precision_full_test = metrics.precision_score(full_y_test, full_test_class)\n",
    "        \n",
    "        if show_fold_stats:        \n",
    "            print('ROC AUC score for XGBoost model valid set: %.4f'%XGB_score)\n",
    "            print('F1 score: %0.4f'%f1s)\n",
    "            print(confusion_matrix(y_val, XGB_class))\n",
    "    \n",
    "            print('ROC AUC score for XGBoost model test set: %.4f'%XGB_test_score)\n",
    "            print('F1 score: %0.4f'%f1_test)\n",
    "            print(confusion_matrix(disease_y_test, XGB_test_class))\n",
    "            \n",
    "            print('ROC AUC score for full test set including other diseases used as false controls: %.4f'%full_test_score)\n",
    "            print('F1 score: %0.4f'%f1_full_test)\n",
    "            print(confusion_matrix(full_y_test, full_test_class))\n",
    "\n",
    "        preds += XGB_test[:,1] / FOLDS\n",
    "        full_preds += full_test_preds[:,1] / FOLDS\n",
    "        \n",
    "        fold_score = [XGB_score,f1s,recall,precision_score, XGB_test_score,f1_test,recall_test,precision_score_test]\n",
    "        oof_scores.append({fold : fold_score})\n",
    "    \n",
    "    avg_test_score = metrics.roc_auc_score(disease_y_test, preds)\n",
    "    avg_class = np.where(preds < 0.5, 0, 1)\n",
    "    avg_f1_test = f1_score(disease_y_test, avg_class)\n",
    "    avg_recall_test = metrics.recall_score(disease_y_test, avg_class)\n",
    "    avg_precision_score_test = metrics.precision_score(disease_y_test, avg_class)\n",
    "\n",
    "    avg_full_test_score = metrics.roc_auc_score(full_y_test, full_preds)\n",
    "    avg_class_full = np.where(full_preds < 0.5, 0, 1)\n",
    "    avg_f1_test_full = f1_score(full_y_test, avg_class_full)\n",
    "    avg_recall_full_test = metrics.recall_score(full_y_test, avg_class_full)\n",
    "    avg_precision_full_test = metrics.precision_score(full_y_test, avg_class_full)\n",
    "\n",
    "    print('-' * 80)\n",
    "    print('Averaged over %i folds ROC AUC score for %s: %.4f'%(FOLDS,d,avg_test_score))\n",
    "    print('F1 : %.4f, Recall : %.4f , Precision : %.4f'%(avg_f1_test, avg_recall_test, avg_precision_score_test))\n",
    "    print('Confusion matrix for %s averaged across %i folds '%(d,FOLDS))\n",
    "    print(confusion_matrix(disease_y_test, avg_class))\n",
    "\n",
    "    \n",
    "    print('Averaged over %i folds ROC AUC score for %s against full set of other diseases: %.4f'%(FOLDS,d,avg_full_test_score))\n",
    "    print('F1 : %.4f, Recall : %.4f , Precision : %.4f'%(avg_f1_test_full, avg_recall_full_test, avg_precision_full_test))\n",
    "    print('Confusion matrix for %s averaged across %i folds for full set'%(d,FOLDS))\n",
    "    print(confusion_matrix(full_y_test, avg_class_full))\n",
    "\n",
    "    if do_plot_ROC:\n",
    "        (fpr, tpr, thresholds) = metrics.roc_curve(disease_y_test, preds)\n",
    "        plot_ROC(fpr, tpr, d)\n",
    "\n",
    "    return preds, disease_y_test, full_preds, full_y_test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.015457,
     "end_time": "2020-12-16T19:23:06.718102",
     "exception": false,
     "start_time": "2020-12-16T19:23:06.702645",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## Predictions\n",
    "We'll use multiple random seeds to have variations with model initialization, and data splits. We will check if ensembling these multiple predictions will give us a better prediction, of if there will be a better seed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-12-16T19:23:06.762419Z",
     "iopub.status.busy": "2020-12-16T19:23:06.761654Z",
     "iopub.status.idle": "2020-12-16T19:26:07.790508Z",
     "shell.execute_reply": "2020-12-16T19:26:07.791135Z"
    },
    "papermill": {
     "duration": 181.058086,
     "end_time": "2020-12-16T19:26:07.791311",
     "exception": false,
     "start_time": "2020-12-16T19:23:06.733225",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "Seed : 42\n",
      "--------------------------------------------------------------------------------\n",
      "Disease : cirrhosis\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9946\n",
      "F1 score: 0.9444\n",
      "[[58  0]\n",
      " [ 2 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9910\n",
      "F1 score: 0.5882\n",
      "[[430  28]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.8707\n",
      "F1 score: 0.6486\n",
      "[[52  6]\n",
      " [ 7 12]]\n",
      "ROC AUC score for XGBoost model test set: 0.9913\n",
      "F1 score: 0.9500\n",
      "[[22  1]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9727\n",
      "F1 score: 0.4810\n",
      "[[418  40]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9807\n",
      "F1 score: 0.9000\n",
      "[[55  2]\n",
      " [ 2 18]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9861\n",
      "F1 score: 0.5067\n",
      "[[422  36]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9860\n",
      "F1 score: 0.8421\n",
      "[[55  2]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9835\n",
      "F1 score: 0.4932\n",
      "[[423  35]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9640\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9872\n",
      "F1 score: 0.5938\n",
      "[[433  25]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis against full set of other diseases: 0.9891\n",
      "F1 : 0.5672, Recall : 0.9500 , Precision : 0.4043\n",
      "Confusion matrix for cirrhosis averaged across 5 folds for full set\n",
      "[[430  28]\n",
      " [  1  19]]\n",
      "================================================================================\n",
      "Seed : 1970\n",
      "--------------------------------------------------------------------------------\n",
      "Disease : cirrhosis\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9936\n",
      "F1 score: 0.9444\n",
      "[[58  0]\n",
      " [ 2 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9922\n",
      "F1 score: 0.6154\n",
      "[[433  25]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9537\n",
      "F1 score: 0.7568\n",
      "[[54  4]\n",
      " [ 5 14]]\n",
      "ROC AUC score for XGBoost model test set: 0.9870\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9640\n",
      "F1 score: 0.4390\n",
      "[[414  44]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9649\n",
      "F1 score: 0.8095\n",
      "[[52  5]\n",
      " [ 3 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9873\n",
      "F1 score: 0.4810\n",
      "[[418  40]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9719\n",
      "F1 score: 0.8205\n",
      "[[54  3]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9821\n",
      "F1 score: 0.5538\n",
      "[[431  27]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9763\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9893\n",
      "F1 score: 0.5846\n",
      "[[432  26]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis against full set of other diseases: 0.9895\n",
      "F1 : 0.5672, Recall : 0.9500 , Precision : 0.4043\n",
      "Confusion matrix for cirrhosis averaged across 5 folds for full set\n",
      "[[430  28]\n",
      " [  1  19]]\n",
      "================================================================================\n",
      "Seed : 1971\n",
      "--------------------------------------------------------------------------------\n",
      "Disease : cirrhosis\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9492\n",
      "F1 score: 0.7500\n",
      "[[57  1]\n",
      " [ 7 12]]\n",
      "ROC AUC score for XGBoost model test set: 0.9935\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9859\n",
      "F1 score: 0.5758\n",
      "[[431  27]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9864\n",
      "F1 score: 0.8824\n",
      "[[58  0]\n",
      " [ 4 15]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9898\n",
      "F1 score: 0.5588\n",
      "[[429  29]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9588\n",
      "F1 score: 0.7895\n",
      "[[54  3]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9835\n",
      "F1 score: 0.5455\n",
      "[[430  28]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9816\n",
      "F1 score: 0.8947\n",
      "[[56  1]\n",
      " [ 3 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9189\n",
      "[[23  0]\n",
      " [ 3 17]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9843\n",
      "F1 score: 0.5484\n",
      "[[433  25]\n",
      " [  3  17]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9724\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9907\n",
      "F1 score: 0.6786\n",
      "[[441  17]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis against full set of other diseases: 0.9904\n",
      "F1 : 0.6129, Recall : 0.9500 , Precision : 0.4524\n",
      "Confusion matrix for cirrhosis averaged across 5 folds for full set\n",
      "[[435  23]\n",
      " [  1  19]]\n",
      "================================================================================\n",
      "Seed : 2020\n",
      "--------------------------------------------------------------------------------\n",
      "Disease : cirrhosis\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9782\n",
      "F1 score: 0.8571\n",
      "[[57  1]\n",
      " [ 4 15]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9824\n",
      "F1 score: 0.5714\n",
      "[[433  25]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9619\n",
      "F1 score: 0.8000\n",
      "[[56  2]\n",
      " [ 5 14]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9876\n",
      "F1 score: 0.5205\n",
      "[[424  34]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9693\n",
      "F1 score: 0.8372\n",
      "[[52  5]\n",
      " [ 2 18]]\n",
      "ROC AUC score for XGBoost model test set: 0.9957\n",
      "F1 score: 0.9500\n",
      "[[22  1]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9912\n",
      "F1 score: 0.5429\n",
      "[[427  31]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9807\n",
      "F1 score: 0.8649\n",
      "[[56  1]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 0.9957\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9811\n",
      "F1 score: 0.6333\n",
      "[[437  21]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9596\n",
      "F1 score: 0.7568\n",
      "[[54  3]\n",
      " [ 6 14]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9873\n",
      "F1 score: 0.5672\n",
      "[[430  28]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis: 0.9978\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis against full set of other diseases: 0.9895\n",
      "F1 : 0.5938, Recall : 0.9500 , Precision : 0.4318\n",
      "Confusion matrix for cirrhosis averaged across 5 folds for full set\n",
      "[[433  25]\n",
      " [  1  19]]\n",
      "================================================================================\n",
      "Seed : 2021\n",
      "--------------------------------------------------------------------------------\n",
      "Disease : cirrhosis\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9410\n",
      "F1 score: 0.7895\n",
      "[[54  4]\n",
      " [ 4 15]]\n",
      "ROC AUC score for XGBoost model test set: 0.9935\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9707\n",
      "F1 score: 0.5000\n",
      "[[424  34]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9701\n",
      "F1 score: 0.8889\n",
      "[[57  1]\n",
      " [ 3 16]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9874\n",
      "F1 score: 0.5714\n",
      "[[428  30]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9623\n",
      "F1 score: 0.6667\n",
      "[[57  0]\n",
      " [10 10]]\n",
      "ROC AUC score for XGBoost model test set: 0.9957\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9856\n",
      "F1 score: 0.6102\n",
      "[[437  21]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9763\n",
      "F1 score: 0.8780\n",
      "[[54  3]\n",
      " [ 2 18]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9912\n",
      "F1 score: 0.7170\n",
      "[[444  14]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9886\n",
      "F1 score: 0.9000\n",
      "[[55  2]\n",
      " [ 2 18]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9889\n",
      "F1 score: 0.6061\n",
      "[[432  26]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 folds ROC AUC score for cirrhosis against full set of other diseases: 0.9913\n",
      "F1 : 0.6129, Recall : 0.9500 , Precision : 0.4524\n",
      "Confusion matrix for cirrhosis averaged across 5 folds for full set\n",
      "[[435  23]\n",
      " [  1  19]]\n"
     ]
    }
   ],
   "source": [
    "SEEDs = [42,1970,1971,2020,2021]        \n",
    "d = 'cirrhosis'\n",
    "d_preds = []\n",
    "df_preds = []\n",
    "dc_preds = []\n",
    "for SEED in SEEDs:\n",
    "    print('=' * 80)\n",
    "    print('Seed : %i'%SEED)\n",
    "    d_pred, y_true, full_preds, full_y_test = disease_prediction(d) \n",
    "    d_preds.append(d_pred)\n",
    "    df_preds.append(full_preds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.025132,
     "end_time": "2020-12-16T19:26:07.841072",
     "exception": false,
     "start_time": "2020-12-16T19:26:07.815940",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "So the model is able to perfectly predict cirrhosis in 'sysnthetic' case when it has to distinguish between healthy samples and samples with the disease. However it gets really confused when we mix 3 unknown diseases to the controls.<br>\n",
    "The best case scenario we get is with seeds 1971 and 2021 giving us F1 = 0.6129 and Precision = 0.4524.<br>\n",
    "<br>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "_kg_hide-input": true,
    "execution": {
     "iopub.execute_input": "2020-12-16T19:26:07.904644Z",
     "iopub.status.busy": "2020-12-16T19:26:07.903490Z",
     "iopub.status.idle": "2020-12-16T19:26:07.925671Z",
     "shell.execute_reply": "2020-12-16T19:26:07.926327Z"
    },
    "papermill": {
     "duration": 0.061153,
     "end_time": "2020-12-16T19:26:07.926574",
     "exception": false,
     "start_time": "2020-12-16T19:26:07.865421",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Averaged predictions across 5 seeds\n",
      "Averaged over 5 seeds ROC AUC score for cirrhosis : 1.0000\n",
      "F1 score : 0.9744\n",
      "Confusion matrix for cirrhosis\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "Averaged over 5 seeds ROC AUC score for cirrhosis against full set of other diseases: 0.9902\n",
      "Confusion matrix for cirrhosis against controls and other diseases combined\n",
      "[[433  25]\n",
      " [  1  19]]\n",
      "F1 score : 0.5938\n",
      "Recall : 0.9500\n",
      "Precision : 0.4318\n"
     ]
    }
   ],
   "source": [
    "print('Averaged predictions across %i seeds'%len(SEEDs))\n",
    "avg_pred = np.array(d_preds).mean(axis= 0)\n",
    "avg_class = np.where(avg_pred < 0.65, 0, 1)\n",
    "print('Averaged over %i seeds ROC AUC score for %s : %.4f'%(len(SEEDs),d, metrics.roc_auc_score(y_true, avg_pred)))\n",
    "print('F1 score : %.4f'%f1_score(y_true, avg_class))\n",
    "print('Confusion matrix for %s'%d)\n",
    "print(confusion_matrix(y_true, avg_class))\n",
    "\n",
    "avg_pred = np.array(df_preds).mean(axis= 0)\n",
    "avg_class = np.where(avg_pred < 0.5, 0, 1)\n",
    "print('Averaged over %i seeds ROC AUC score for %s against full set of other diseases: %.4f'%(len(SEEDs),d, metrics.roc_auc_score(full_y_test, avg_pred)))\n",
    "print('Confusion matrix for %s against controls and other diseases combined'%d)\n",
    "print(confusion_matrix(full_y_test, avg_class))\n",
    "print('F1 score : %.4f'%f1_score(full_y_test, avg_class))\n",
    "print('Recall : %.4f'%metrics.recall_score(full_y_test, avg_class))\n",
    "print('Precision : %.4f'%metrics.precision_score(full_y_test, avg_class))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.024153,
     "end_time": "2020-12-16T19:26:07.975550",
     "exception": false,
     "start_time": "2020-12-16T19:26:07.951397",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Averaging results across all variants does not improve prediction in this case, model with the seed 2021 is still the best. <br>\n",
    "We need significantly less correlated models in the ensemble for it to enhance the prediction.<br>\n",
    "<br>\n",
    "However if we adjust threshold of class separation to 0.88 then we get much better F1 and Precision by virtually eliminating false positives (and increasing number of false negatives at the same time, with Recall worsening as well). So as usual it comes to the decision of what is more important - Precision or Recall."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "_kg_hide-input": true,
    "execution": {
     "iopub.execute_input": "2020-12-16T19:26:08.035213Z",
     "iopub.status.busy": "2020-12-16T19:26:08.034116Z",
     "iopub.status.idle": "2020-12-16T19:26:08.047633Z",
     "shell.execute_reply": "2020-12-16T19:26:08.048392Z"
    },
    "papermill": {
     "duration": 0.047727,
     "end_time": "2020-12-16T19:26:08.048579",
     "exception": false,
     "start_time": "2020-12-16T19:26:08.000852",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Averaged over 5 seeds ROC AUC score for cirrhosis against full set of other diseases: 0.9902\n",
      "Confusion matrix for cirrhosis against controls and other diseases combined\n",
      "[[456   2]\n",
      " [  5  15]]\n",
      "F1 score : 0.8108\n",
      "Recall : 0.7500\n",
      "Precision : 0.8824\n"
     ]
    }
   ],
   "source": [
    "avg_class = np.where(avg_pred < 0.88, 0, 1)\n",
    "print('Averaged over %i seeds ROC AUC score for %s against full set of other diseases: %.4f'%(len(SEEDs),d, metrics.roc_auc_score(full_y_test, avg_pred)))\n",
    "print('Confusion matrix for %s against controls and other diseases combined'%d)\n",
    "print(confusion_matrix(full_y_test, avg_class))\n",
    "print('F1 score : %.4f'%f1_score(full_y_test, avg_class))\n",
    "print('Recall : %.4f'%metrics.recall_score(full_y_test, avg_class))\n",
    "print('Precision : %.4f'%metrics.precision_score(full_y_test, avg_class))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.024647,
     "end_time": "2020-12-16T19:26:08.098992",
     "exception": false,
     "start_time": "2020-12-16T19:26:08.074345",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "## Discriminate cirrhosis from other diseases in pairs\n",
    "As it was discovered in the original paper diabetes is the most difficult to predict based off the abundance features, and I suspect that is what prohibiting the model to discriminate cirrhosis from other samples, and generates false positives. And we are going to check it in the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2020-12-16T19:26:08.164668Z",
     "iopub.status.busy": "2020-12-16T19:26:08.162819Z",
     "iopub.status.idle": "2020-12-16T19:28:04.412313Z",
     "shell.execute_reply": "2020-12-16T19:28:04.411103Z"
    },
    "papermill": {
     "duration": 116.288271,
     "end_time": "2020-12-16T19:28:04.412675",
     "exception": false,
     "start_time": "2020-12-16T19:26:08.124404",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================================================================================\n",
      "Seed : 1970\n",
      "--------------------------------------------------------------------------------\n",
      "Discriminate cirrhosis from obesity\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9936\n",
      "F1 score: 0.9444\n",
      "[[58  0]\n",
      " [ 2 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 1.0000\n",
      "F1 score: 0.9756\n",
      "[[186   1]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9537\n",
      "F1 score: 0.7568\n",
      "[[54  4]\n",
      " [ 5 14]]\n",
      "ROC AUC score for XGBoost model test set: 0.9870\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9880\n",
      "F1 score: 0.8182\n",
      "[[181   6]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9649\n",
      "F1 score: 0.8095\n",
      "[[52  5]\n",
      " [ 3 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9989\n",
      "F1 score: 0.9500\n",
      "[[186   1]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9719\n",
      "F1 score: 0.8205\n",
      "[[54  3]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9979\n",
      "F1 score: 0.9231\n",
      "[[186   1]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9763\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9992\n",
      "F1 score: 0.9500\n",
      "[[186   1]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "ROC AUC score for cirrhosis averaged over 5 folds: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for cirrhosis against obesity averaged over 5 folds : 0.9995\n",
      "F1 : 0.9500, Recall : 0.9500 , Precision : 0.9500\n",
      "Confusion matrix for cirrhosis against obesity averaged across 5 folds\n",
      "[[186   1]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Discriminate cirrhosis from t2d\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9936\n",
      "F1 score: 0.9444\n",
      "[[58  0]\n",
      " [ 2 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9896\n",
      "F1 score: 0.6780\n",
      "[[227  19]\n",
      " [  0  20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9537\n",
      "F1 score: 0.7568\n",
      "[[54  4]\n",
      " [ 5 14]]\n",
      "ROC AUC score for XGBoost model test set: 0.9870\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9518\n",
      "F1 score: 0.5143\n",
      "[[214  32]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9649\n",
      "F1 score: 0.8095\n",
      "[[52  5]\n",
      " [ 3 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9809\n",
      "F1 score: 0.5135\n",
      "[[211  35]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9719\n",
      "F1 score: 0.8205\n",
      "[[54  3]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9740\n",
      "F1 score: 0.6102\n",
      "[[225  21]\n",
      " [  2  18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9763\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9854\n",
      "F1 score: 0.6333\n",
      "[[225  21]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "ROC AUC score for cirrhosis averaged over 5 folds: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for cirrhosis against t2d averaged over 5 folds : 0.9854\n",
      "F1 : 0.6333, Recall : 0.9500 , Precision : 0.4750\n",
      "Confusion matrix for cirrhosis against t2d averaged across 5 folds\n",
      "[[225  21]\n",
      " [  1  19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Discriminate cirrhosis from cancer\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 1\n",
      "ROC AUC score for XGBoost model valid set: 0.9936\n",
      "F1 score: 0.9444\n",
      "[[58  0]\n",
      " [ 2 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 1.0000\n",
      "[[23  0]\n",
      " [ 0 20]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9859\n",
      "F1 score: 0.8889\n",
      "[[66  5]\n",
      " [ 0 20]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 2\n",
      "ROC AUC score for XGBoost model valid set: 0.9537\n",
      "F1 score: 0.7568\n",
      "[[54  4]\n",
      " [ 5 14]]\n",
      "ROC AUC score for XGBoost model test set: 0.9870\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9577\n",
      "F1 score: 0.8182\n",
      "[[65  6]\n",
      " [ 2 18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 3\n",
      "ROC AUC score for XGBoost model valid set: 0.9649\n",
      "F1 score: 0.8095\n",
      "[[52  5]\n",
      " [ 3 17]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9873\n",
      "F1 score: 0.8837\n",
      "[[67  4]\n",
      " [ 1 19]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 4\n",
      "ROC AUC score for XGBoost model valid set: 0.9719\n",
      "F1 score: 0.8205\n",
      "[[54  3]\n",
      " [ 4 16]]\n",
      "ROC AUC score for XGBoost model test set: 0.9978\n",
      "F1 score: 0.9474\n",
      "[[23  0]\n",
      " [ 2 18]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9789\n",
      "F1 score: 0.8372\n",
      "[[66  5]\n",
      " [ 2 18]]\n",
      "--------------------------------------------------------------------------------\n",
      "Fold : 5\n",
      "ROC AUC score for XGBoost model valid set: 0.9763\n",
      "F1 score: 0.8333\n",
      "[[56  1]\n",
      " [ 5 15]]\n",
      "ROC AUC score for XGBoost model test set: 1.0000\n",
      "F1 score: 0.9744\n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for full test set including other diseases used as false controls: 0.9838\n",
      "F1 score: 0.8837\n",
      "[[67  4]\n",
      " [ 1 19]]\n",
      "--------------------------------------------------------------------------------\n",
      "ROC AUC score for cirrhosis averaged over 5 folds: 1.0000\n",
      "F1 : 0.9744, Recall : 0.9500 , Precision : 1.0000\n",
      "Confusion matrix for cirrhosis averaged across 5 folds \n",
      "[[23  0]\n",
      " [ 1 19]]\n",
      "ROC AUC score for cirrhosis against cancer averaged over 5 folds : 0.9845\n",
      "F1 : 0.8444, Recall : 0.9500 , Precision : 0.7600\n",
      "Confusion matrix for cirrhosis against cancer averaged across 5 folds\n",
      "[[65  6]\n",
      " [ 1 19]]\n"
     ]
    }
   ],
   "source": [
    "dc_preds = []\n",
    "SEED = 1970\n",
    "print('=' * 80)\n",
    "print('Seed : %i'%SEED)\n",
    "\n",
    "for counter in diseases:\n",
    "    if counter != d:\n",
    "        dc_pred, y_preds, _ = disease_distinction(d, counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.031142,
     "end_time": "2020-12-16T19:28:04.476312",
     "exception": false,
     "start_time": "2020-12-16T19:28:04.445170",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "As we see the model is capable to distinguish cirrhosis from obesity and cancer with high precision. However diabetes leads to a significant number of false positives, lowering precision to about 50%.<br>\n",
    "We will need to work on feature engineering to improve diabetes prediction. There is hope in usage of markers instead of/in conjunction with abundance features will help, as it's been shown in the original paper.<br>\n",
    "And of course there is room for improvement in the XGB model's hyper-parameters tuning, as I've used it with basic default parameters only yet."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "papermill": {
     "duration": 0.031011,
     "end_time": "2020-12-16T19:28:04.538956",
     "exception": false,
     "start_time": "2020-12-16T19:28:04.507945",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Conclusion\n",
    "Using XGBoost defintely provides significant boost in precision of predicting cirrhosis in comparison with RF results published in the paper. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "papermill": {
   "duration": 308.702954,
   "end_time": "2020-12-16T19:28:04.679163",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2020-12-16T19:22:55.976209",
   "version": "2.1.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
